### ISSUE TITLE
How to change identity provider to htpasswd in oc 3.11

### ISSUE THREAD

**Issue by EnilPajic** [openshift/origin#24566]
It is not an issue per-se, but a question. How can I change identity provider to `HTPasswdPasswordIdentityProvider`?.  I'm using origin version 3.11. In 3.9 version I would just change the `master-config.yaml` file, key `identityProviders`, to something like this:

```yaml
oauthConfig:
  alwaysShowProviderSelection: false
  assetPublicURL: https://X.Y.Z.W:8443/console/
  grantConfig:
    method: auto
    serviceAccountMethod: prompt
  identityProviders:
  - challenge: true
    login: true
    mappingMethod: claim
    name: my_htpasswd_provider
    provider:
      apiVersion: v1
      kind: HTPasswdPasswordIdentityProvider
      file: /root/openshift/data-311/openshift-controller-manager/users.htpasswd 
  masterCA: ca-bundle.crt
  masterPublicURL: https://X.Y.Z.W:8443
  masterURL: https://X.Y.Z.W:8443
  sessionConfig:
    sessionMaxAgeSeconds: 300
    sessionName: ssn
    sessionSecretsFile: ""
  templates: null
  tokenConfig:
    accessTokenMaxAgeSeconds: 86400
    authorizeTokenMaxAgeSeconds: 300
```

In version 3.11 this is not working because there is no `master` directory in openshift. The file `master-config.yaml` is present in 3 directories: `openshift-controller-manager`, `openshift-apiserver`, `openshift-controller-manager`. And I don't know which one to change. 

[Documentation](https://docs.openshift.com/container-platform/3.11/install_config/configuring_authentication.html#identity-providers_parameters) is always referring to "default master config location" which is `/etc/origin/master/master-config.yaml`, but in that folder there is just an empty folder `cloudproviders`.

By "not working" I mean oc continues to use `AllowAllPasswordIdentityProvider` when I change `master-config.yaml` in `openshift-controller-manager` (after cluster restart, of course) . When I change `master-config.yaml` in two other places, or only one of them (`openshift-apiserver`, `kube-apiserver`), then cluster won't start. Error is something like "cluster health check failed". I assume this is because some endpoints were not supposed to be auth-protected.

Tried to change `mappingMethod` to `add/lookup`, without success.
Also tried to move `users.htpasswd` file outside openshift's folders.

I've installed oc on Ubuntu (downloaded oc tools), server info:
```
18.04.1 LTS (Bionic Beaver)
Linux dev-staging-de-fra 4.15.0-32-generic #35-Ubuntu SMP Fri Aug 10 17:58:07 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

There are two "workarounds" to this problem, neither satisfying enough:
1. use oc 3.9 (I would like to use 3.11)
2. set `masterPublicURL` to localhost and access server via ssh tunnel, so no one can log into web console without accessing the server itself (this is not acceptable solution because I want to add team members to some oc projects and not give them server access)

##### Version
oc v3.11.0+0cbc58b
kubernetes v1.11.0+d4cacc0
features: Basic-Auth GSSAPI Kerberos SPNEGO

Server https://X.Y.Z.W:8443
kubernetes v1.11.0+d4cacc0

##### Steps To Reproduce
1. download oc tools: `wget https://github.com/openshift/origin/releases/download/v3.11.0/openshift-origin-client-tools-v3.11.0-0cbc58b-linux-64bit.tar.gz`
2. unpack (tar xvf), move oc to /usr/local/bin
3. start cluster with custom base dir: 
`oc cluster up --write-config=false --server-loglevel=5 --base-dir=/root/openshift/data-311 --public-hostname=X.Y.Z.W`
4. change `master-config.yaml` in `openshift-data/openshift-controller-manager` with Htpasswd identity provider.
5. restart cluster
6. oc login (enter any name as username and any password)

(on a side note, when I type `--write-config=true` cluster wont start at all)

##### Current Result
User is created and logged in
##### Expected Result
To not log in if there is no user

##### Additional Information
There is one error when I run diagnostics: cannot log with admin user (`system:admin`), maybe because I changed IdentityProvider (I cannot log in with `oc login -u system:admin`). I ran the diagnostics with default `developer` user. There are other users (`abc`, `d`, `ds`) that were created while I tried to log in with "wrong" username/password.

```
oc adm diagnostics
[Note] Determining if client configuration exists for client/cluster diagnostics
Info:  Successfully read a client config file at '/root/.kube/config'
[Note] Could not configure a client with cluster-admin permissions for the current server, so cluster diagnostics will be skipped

[Note] Running diagnostic: ConfigContexts[default/X-Y-Z-W:8443/system:admin]
       Description: Validate client config context is complete and has connectivity
       
ERROR: [DCli0015 from diagnostic ConfigContexts@openshift/origin/pkg/oc/cli/admin/diagnostics/diagnostics/client/config_contexts.go:298]
       For client config context 'default/X-Y-Z-W:8443/system:admin':
       The server URL is 'https://X.Y.Z.W:8443'
       The user authentication is 'system:admin/X-Y-Z-W:8443'
       The current project is 'default'
       (*errors.StatusError) Unauthorized
       Diagnostics does not have an explanation for what this means. Please report this error so one can be added.
       
[Note] Running diagnostic: ConfigContexts[myproject/X-Y-Z-W:8443/developer]
       Description: Validate client config context is complete and has connectivity
       
Info:  The current client config context is 'myproject/X-Y-Z-W:8443/developer':
       The server URL is 'https://X.Y.Z.W:8443'
       The user authentication is 'developer/X-Y-Z-W:8443'
       The current project is 'myproject'
       Successfully requested project list; has access to project(s):
         [myproject]
       
[Note] Running diagnostic: ConfigContexts[/X-Y-Z-W:8443/abc]
       Description: Validate client config context is complete and has connectivity
       
Info:  For client config context '/X-Y-Z-W:8443/abc':
       The server URL is 'https://X.Y.Z.W:8443'
       The user authentication is 'abc/X-Y-Z-W:8443'
       The current project is 'default'
       Successfully requested project list, but it is empty, so user has no access to anything.
       
[Note] Running diagnostic: ConfigContexts[/X-Y-Z-W:8443/d]
       Description: Validate client config context is complete and has connectivity
       
Info:  For client config context '/X-Y-Z-W:8443/d':
       The server URL is 'https://X.Y.Z.W:8443'
       The user authentication is 'd/X-Y-Z-Q:8443'
       The current project is 'default'
       Successfully requested project list, but it is empty, so user has no access to anything.
       
[Note] Running diagnostic: ConfigContexts[/X-Y-Z-W:8443/ds]
       Description: Validate client config context is complete and has connectivity
       
Info:  For client config context '/X-Y-Z-W:8443/ds':
       The server URL is 'https://X.Y.Z.W8443'
       The user authentication is 'ds/X-Y-Z-W:8443'
       The current project is 'default'
       Successfully requested project list, but it is empty, so user has no access to anything.
       
[Note] Running diagnostic: DiagnosticPod
       Description: Create a pod to run diagnostics from the application standpoint
       
Info:  Output from the diagnostic pod (image openshift/origin-deployer:v3.11.0):
       [Note] Running diagnostic: PodCheckAuth
              Description: Check that service account credentials authenticate as expected
              
       Info:  Service account token successfully authenticated to master
       Info:  Service account token was authenticated by the integrated registry.
       
       [Note] Running diagnostic: PodCheckDns
              Description: Check that DNS within a pod works as expected
              
       [Note] Summary of diagnostics execution (version v3.11.0+4986039-380):
       [Note] Completed with no errors or warnings seen.
       
[Note] Summary of diagnostics execution (version v3.11.0+0cbc58b):
[Note] Errors seen: 1
```

**Comment by mutaztom** [comment]
I have the same problem, I have okd installed in Centos 7.

**Comment by raesene** [comment]
so I managed to sort this out, so I can provide one approach which worked for me (YMMV of course).  Using the oc tool standalone on a Centos7 VM running in EU-West-2, here's the steps I took to enable the HTPasswd provider.

1. Bring up the cluster `oc cluster up --base-dir=/home/centos/osconfig/ --routing-suffix=[Centos.SSH.IP].nip.io --public-hostname=ec2-[centos-ssh-ip].eu-west-2.compute.amazonaws.com`

2. Install the tools `sudo yum install httpd-tools`

3. Create a user called admin `htpasswd -c /home/centos/osconfig/kube-apiserver/htpasswd admin`

4. Add admin to `cluster-admin` (fine in test don't do this in prod!) `oc create clusterrolebinding admin --clusterrole=cluster-admin --user=admin`

5. Modify the master-config.yaml file `nano /home/centos/osconfig/kube-apiserver/master-config.yaml` . replace the identity provider section with

```
  identityProviders:
  - name: my_htpasswd_provider 
    challenge: true 
    login: true 
    mappingMethod: claim 
    provider:
      apiVersion: v1
      kind: HTPasswdPasswordIdentityProvider
      file: /etc/origin/master/htpasswd
```

6. stop the cluster `oc cluster down`

7. start the cluster `oc cluster up --base-dir=/home/centos/osconfig/ --routing-suffix=[Centos.SSH.IP].nip.io --public-hostname=ec2-[centos-ssh-ip].eu-west-2.compute.amazonaws.com`

That should work :)
